torch/csrc/jit/codegen/cuda/fusion_segmenter.h

```diff
@@ -0,0 +1,365 @@
+  std::string toString(int verbosity) const;
+
+  // Debug print for segmented fusions
+  void print();
+
+  //! API for adding groups
+  SegmentedGroup* newGroup();
+
+  //! API shortcut for adding a singleton group
+  SegmentedGroup* newGroup(Expr* expr);
+
+  //! API for adding edges
+  SegmentedEdge* newEdge(SegmentedGroup* from, SegmentedGroup* to, Val* val);
+
+ protected:
+  //! original full fusion
+  Fusion fusion_;
+
+  //! Count total exprs
+  size_t total_expr_count_ = 0;
+
+  //! States representing segmentation
+  std::vector<SegmentedEdge*> edges_;
+  std::vector<SegmentedGroup*> groups_;
+
+  // Owning object to explicitly manage groups and edges
+  // Owning object to provide
+  class Impl {
+   public:
+    explicit Impl(SegmentedFusion* sf) : owning_fusion_(sf) {}
+
+    SegmentedGroup* makeGroup();
+    SegmentedGroup* makeGroup(Expr*);
+    SegmentedEdge* makeEdge(SegmentedGroup* from, SegmentedGroup* to, Val* val);
+    void cleanUnused();
+
+   private:
+    using GroupPtr = std::unique_ptr<SegmentedGroup>;
+    using EdgePtr = std::unique_ptr<SegmentedEdge>;
+    std::vector<GroupPtr> groups_;
+    std::vector<EdgePtr> edges_;
+    SegmentedFusion* owning_fusion_;
+  };
+  Impl impl_;
+
+ protected:
+  friend class SegmentCandidateFinder;
+  //! Make a heuristics entry for a group and parameters
+  std::unique_ptr<SchedulerEntry> makeSchedulerEntry(
+      SegmentedGroup* sg,
+      ExpressionEvaluator& ee);
+
+  //! Cleanup function to be call at the end of fusion
+  //!  segment pass
+  void finalize();
+};
+
+class TORCH_CUDA_API SegmentCandidateFinder {
+ public:
+  // Take a copy of fusion to own
+  SegmentCandidateFinder(const Fusion* fusion);
+
+  static std::unique_ptr<SegmentedFusion> segment(const Fusion* fusion) {
+    SegmentCandidateFinder scf(fusion);
+    return std::move(scf.segmented_fusion);
+  }
+
+ private:
+  void resetTraversal();
+
+  void resetLevels();
+
+  void mergeNodes();
+
+  bool codeGenSupportedMerge(SegmentedEdge* edge);
+
+  void findSegments();
+
+  std::unordered_set<SegmentedEdge*> disconnectGroup(SegmentedGroup* group);
+
+  std::vector<SegmentedGroup*>& groups() {
+    TORCH_INTERNAL_ASSERT(
+        segmented_fusion != nullptr, "Segment finder not owinging any fusion");
+    return segmented_fusion->groups();
+  }
+
+  std::vector<SegmentedEdge*>& edges() {
+    TORCH_INTERNAL_ASSERT(
+        segmented_fusion != nullptr, "Segment finder not owinging any fusion");
+    return segmented_fusion->edges();
+  }
+
+  Fusion& completeFusion() {
+    TORCH_INTERNAL_ASSERT(
+        segmented_fusion != nullptr, "Segment finder not owinging any fusion");
+    return segmented_fusion->completeFusion();
+  }
+
+  void finalize();
+
+  // Return the resulting heuristic corresponding to the merged
+  //  group built by merging the two groups connected by edge
+  ScheduleHeuristic deriveHeuristic(SegmentedGroup* edge);
+
+ protected:
+  std::deque<SegmentedGroup*> to_visit;
```

Comment:
private & protected data members should have _ suffix...

Reply:
Updated, thanks!